numerics/next.hpp

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <limits>
+#include <type_traits>
+
+#include "numerics/log_b.hpp"
+#include "numerics/scale_b.hpp"
+
+namespace principia {
+namespace numerics {
+
+// A constexpr implementation of the IEEE 754:2008 nextUp and nextDown
+// functions.
+template<typename SourceFormat,
+         typename = std::enable_if_t<std::is_floating_point_v<SourceFormat>>>
+constexpr SourceFormat NextUp(SourceFormat const x) {
+  // Handle the edge cases in the order in which they are listed in
+  // IEEE 754:2008.
+  if (x == -std::numeric_limits<SourceFormat>::denorm_min()) {
+    return -0.0;
+  }
+  if (x == 0) {
+    // Note that this is independent of the sign of 0, as specified by
+    // IEEE 754:2008, 5.3.1.  This differs from C++ std::nextafter, which maps
+    // -0 to +0.
+    return std::numeric_limits<SourceFormat>::denorm_min();
+  }
+  if (x == std::numeric_limits<SourceFormat>::infinity()) {
+    // TODO(egg): Should we be signalling the overflow exception?
```

Comment:
I don't think we ever use the floating-point exceptions...

Reply:
No, but when it’s trivial to get them right we might as well…